RecFCCeeCalorimeter/src/components/CaloTowerToolFCCee.cpp

```diff
@@ -0,0 +1,431 @@
+  // tower index of the borders of the cell
+  int iTheta = 0;
+  int iPhi = 0;
+  const dd4hep::DDSegmentation::FCCSWGridModuleThetaMerged* segmentation = nullptr;
+  const dd4hep::DDSegmentation::MultiSegmentation* multisegmentation = nullptr;
+  if( aType == SegmentationType::kModuleTheta) {
+      segmentation = dynamic_cast<const dd4hep::DDSegmentation::FCCSWGridModuleThetaMerged*>(aSegmentation);
+  } else if( aType == SegmentationType::kMulti) {
+    multisegmentation = dynamic_cast<const dd4hep::DDSegmentation::MultiSegmentation*>(aSegmentation);
+  }
+  bool pass = true;
+  for (const auto& cell : *aCells) {
+    pass = true;
+    // if multisegmentation is used - first find out which segmentation to use
+    if( aType == SegmentationType::kMulti) {
+      segmentation = dynamic_cast<const dd4hep::DDSegmentation::FCCSWGridModuleThetaMerged*>(&multisegmentation->subsegmentation(cell.getCellID()));
+    }
+    if (m_useHalfTower) {
+        uint layerId = m_decoder->get(cell.getCellID(), "layer");
+        if ( layerId > m_max_layer ) {
+          pass = false;
+        }
+    }
+    if (pass) {
+      float cellX = cell.getPosition().x;
+      float cellY = cell.getPosition().y;
+      float cellZ = cell.getPosition().z;
+      float cellTheta = atan2(sqrt(cellX * cellX + cellY * cellY), cellZ);
+      float cellPhi   = atan2(cellY, cellX);
+      iTheta = idTheta(cellTheta);
+      iPhi = idPhi(cellPhi);
+      aTowers[iTheta][phiNeighbour(iPhi)] +=
+        cell.getEnergy() * sin(segmentation->theta(cell.getCellID()));
+      if (fillTowersCells) {
+        m_cellsInTowers[std::make_pair(iTheta, phiNeighbour(iPhi))].push_back(cell.clone());
+        if ( m_cellsInTowers[std::make_pair(iTheta, phiNeighbour(iPhi))].size() > 5 )
+          verbose() << "NUM CELLs IN TOWER : " << m_cellsInTowers[std::make_pair(iTheta, phiNeighbour(iPhi))].size() << endmsg;
+      }
+    }
+  }
+}
+
+std::pair<dd4hep::DDSegmentation::Segmentation*, CaloTowerToolFCCee::SegmentationType> CaloTowerToolFCCee::retrieveSegmentation(std::string aReadoutName) {
+  dd4hep::DDSegmentation::Segmentation* segmentation = nullptr;
+  if (m_geoSvc->lcdd()->readouts().find(aReadoutName) == m_geoSvc->lcdd()->readouts().end()) {
+    info() << "Readout does not exist! Please check if it is correct. Processing without it." << endmsg;
+  } else {
+    info() << "Readout " << aReadoutName << " found." << endmsg;
+    segmentation = dynamic_cast<dd4hep::DDSegmentation::FCCSWGridModuleThetaMerged*>(
+      m_geoSvc->lcdd()->readout(aReadoutName).segmentation().segmentation());
+    if (segmentation == nullptr) {
+      segmentation = dynamic_cast<dd4hep::DDSegmentation::MultiSegmentation*>(
+        m_geoSvc->lcdd()->readout(aReadoutName).segmentation().segmentation());
+      if (segmentation == nullptr) {
+        error() << "There is no module-theta or multi- segmentation for the readout " << aReadoutName << " defined." << endmsg;
```

Comment:
Can you return something like `return std::make_pair(nullptr, SegmentationType::kWrong);` and than handle it, so the algorithm crashes nicely?

Reply:
OK.